README.md
# sublime-edit-command-palette

Add/edit commands for your Command Palette

This plugin was created out of frustration that multiple plugin creators write key bindings but nothing for the command palette.

![Animation](https://cloud.githubusercontent.com/assets/902488/4873073/c09b61c2-6200-11e4-8d1c-f149558ef04c.gif)

## Getting Started
### Installation
This package is available under `Edit Command Palette` inside of [Package Control][], a [Sublime Text][] plugin that allows for easy management of other plugins.

[Sublime Text]: http://www.sublimetext.com/
[Package Control]: http://wbond.net/sublime_packages/package_control

If you prefer the manual route, you can install the script via the following command in the Sublime Text terminal (``ctrl+` ``) which utilizes `git clone`.

```python
import os; path=sublime.packages_path(); (os.makedirs(path) if not os.path.exists(path) else None); window.run_command('exec', {'cmd': ['git', 'clone', 'https://github.com/twolfson/sublime-edit-command-palette', 'Edit Command Palette'], 'working_dir': path})
```

Packages can be uninstalled via "Package Control: Remove Package" via the command pallete, `ctrl+shift+p` on Windows/Linux, `command+shift+p` on Mac.

### Command Palette
`sublime-edit-command-palette` provides 2 commands to access Sublime Text's commands:

- `Preferences: Commands - Default` - Opens default commands provided by Sublime Text
- `Preferences: Commands - User` - Opens custom user commands and provides prompt for first-timers

### Menu
`sublime-edit-command-palette` defines 2 menu items under `Preferences`:

- `Commands - Default` - Opens default commands provided by Sublime Text
- `Commands - User` - Opens custom user commands and provides prompt for first-timers

## Documentation
### `edit_command_palette_open_user(self)`
`edit_command_palette_open_user` creates a default `Packages/User/Default.sublime-commands` containing a prompt for the user, opens it, and sets the language to JSON.

We add in the default and language to make it more approachable. Otherwise, it would be an empty file and in plain text by default.

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality.

## Donating
Support this project and [others by twolfson][gratipay] via [gratipay][].

[![Support via Gratipay][gratipay-badge]][gratipay]

[gratipay-badge]: https://cdn.rawgit.com/gratipay/gratipay-badge/2.x.x/dist/gratipay.png
[gratipay]: https://www.gratipay.com/twolfson/

## Unlicense
As of Oct 31 2014, Todd Wolfson has released this repository and its contents to the public domain.

It has been released under the [UNLICENSE][].

[UNLICENSE]: UNLICENSE


CHANGELOG.md
# sublime-edit-command-palette changelog
2.1.0 - Repaired opening User commands on Windows. Fixed #2

2.0.1 - Updated `install.md` to have `README.md's` rename changes

2.0.0 - Renamed from `sublime-commands` to `sublime-edit-command-palette` to appease Package Control maintainers

1.3.0 - Added `commands_open_file` to open a commands file with desired syntax

1.2.0 - Moved `default_prompt.json` into plugin to avoid ST3 packaging issues

1.1.0 - Fixed up Windows, ST3, and documentation issues for Package Control submission

1.0.0 - Initial release


commands.py
from os import path
import sublime
import sublime_plugin

# Define dictionary for path constants and default content
C = {}
DEFAULT_CONTENT = """[
    // Define your custom Sublime commands here
    // Format is same as `key bindings` except replace `keys` with `caption`
    // {
    //     // Name to show in Command Palette
    //     "caption": "File: New",
    //     // Command to invoke
    //     "command": "new_file",
    //     // Optional keyword arguments to provide to command
    //     "args": {"key": "value"}
    // }
]"""


def plugin_loaded():
    """Once the plugin has loaded, define constants"""
    C['SUBLIME_ROOT'] = path.normpath(path.join(sublime.packages_path(), '..'))
    C['COMMANDS_FILEPATH'] = path.join('Packages', 'User', 'Default.sublime-commands')
    C['COMMANDS_FULL_FILEPATH'] = path.join(C['SUBLIME_ROOT'], C['COMMANDS_FILEPATH'])


class EditCommandPaletteOpenFileCommand(sublime_plugin.WindowCommand):
    def run(self, file):
        """Open a commands file with default JSON syntax"""
        # Open the commands file
        # DEV: We use this because `${packages}` is not easy to resolve in ST3
        self.window.run_command('open_file', {
            'file': file,
        })

        # If the syntax is plain text, move to JSON
        # DEV: Syntax paths always use `/`, even on Windows via https://github.com/wbond/package_control_channel/pull/3780#issuecomment-61369387
        view = self.window.active_view()
        if view.settings().get('syntax') == 'Packages/Text/Plain text.tmLanguage':
            view.set_syntax_file('Packages/JavaScript/JSON.tmLanguage')


class EditCommandPaletteOpenUserCommand(sublime_plugin.WindowCommand):
    def run(self):
        """Open `Packages/User/Default.sublime-commands` for custom definitions"""
        # If the User commands doesn't exist, provide a prompt
        filepath = C['COMMANDS_FULL_FILEPATH']
        if not path.exists(filepath):
            with open(filepath, 'w') as f:
                f.write(DEFAULT_CONTENT)

        # Open the commands file
        self.window.run_command('edit_command_palette_open_file', {
            'file': filepath,
        })


# If we are not in ST3, run `plugin_loaded` by default
if sublime.version() < '3000':
    plugin_loaded()


edit_command_palette.py
from os import path
import sublime
import sublime_plugin

# Define dictionary for path constants and default content
C = {}
DEFAULT_CONTENT = """[
    // Define your custom Sublime commands here
    // Format is same as `key bindings` except replace `keys` with `caption`
    // {
    //     // Name to show in Command Palette
    //     "caption": "File: New",
    //     // Command to invoke
    //     "command": "new_file",
    //     // Optional keyword arguments to provide to command
    //     "args": {"key": "value"}
    // }
]"""


def plugin_loaded():
    """Once the plugin has loaded, define constants"""
    C['SUBLIME_ROOT'] = path.normpath(path.join(sublime.packages_path(), '..'))
    C['COMMANDS_FILEPATH'] = path.join('Packages', 'User', 'Default.sublime-commands')
    C['COMMANDS_SUBLIME_FILEPATH'] = '${packages}/User/Default.sublime-commands'
    C['COMMANDS_FULL_FILEPATH'] = path.join(C['SUBLIME_ROOT'], C['COMMANDS_FILEPATH'])


class EditCommandPaletteOpenFileCommand(sublime_plugin.WindowCommand):
    def run(self, file):
        """Open a commands file with default JSON syntax"""
        # Open the commands file
        # DEV: We use this because `${packages}` is not easy to resolve in ST3
        self.window.run_command('open_file', {
            'file': file,
        })

        # If the syntax is plain text, move to JSON
        # DEV: Syntax paths always use `/`, even on Windows via https://github.com/wbond/package_control_channel/pull/3780#issuecomment-61369387
        view = self.window.active_view()
        if view.settings().get('syntax') == 'Packages/Text/Plain text.tmLanguage':
            view.set_syntax_file('Packages/JavaScript/JSON.tmLanguage')


class EditCommandPaletteOpenUserCommand(sublime_plugin.WindowCommand):
    def run(self):
        """Open `Packages/User/Default.sublime-commands` for custom definitions"""
        # If the User commands doesn't exist, provide a prompt
        filepath = C['COMMANDS_FULL_FILEPATH']
        if not path.exists(filepath):
            with open(filepath, 'w') as f:
                f.write(DEFAULT_CONTENT)

        # Open the commands file
        self.window.run_command('edit_command_palette_open_file', {
            'file': C['COMMANDS_SUBLIME_FILEPATH'],
        })


# If we are not in ST3, run `plugin_loaded` by default
if sublime.version() < '3000':
    plugin_loaded()
